pkg/virtual/syncer/builder/build.go

```diff
@@ -0,0 +1,179 @@
+			}
+			completedContext = requestContext
+			if !strings.HasPrefix(urlPath, rootPathPrefix) {
+				return
+			}
+			withoutRootPathPrefix := strings.TrimPrefix(urlPath, rootPathPrefix)
+			parts := strings.SplitN(withoutRootPathPrefix, "/", 3)
+			if len(parts) < 2 || parts[0] == "" || parts[1] == "" {
+				return
+			}
+			workloadClusterKey := syncer.WorkloadClusterRef{
+				LogicalClusterName: logicalcluster.New(parts[0]),
+				Name:               parts[1],
+			}.Key()
+
+			realPath := "/"
+			if len(parts) > 2 {
+				realPath += parts[2]
+			}
+
+			cluster := genericapirequest.Cluster{Name: logicalcluster.Wildcard, Wildcard: true}
+			if strings.HasPrefix(realPath, "/clusters/") {
+				withoutClustersPrefix := strings.TrimPrefix(realPath, "/clusters/")
+				parts = strings.SplitN(withoutClustersPrefix, "/", 2)
+				lclusterName := parts[0]
+				realPath = "/"
+				if len(parts) > 1 {
+					realPath += parts[1]
+				}
+				cluster = genericapirequest.Cluster{Name: logicalcluster.New(lclusterName)}
+				if lclusterName == "*" {
+					cluster.Wildcard = true
+				}
+			}
+
+			completedContext = dynamiccontext.WithAPIDomainKey(
+				genericapirequest.WithCluster(requestContext, cluster),
+				workloadClusterKey)
+			prefixToStrip = strings.TrimSuffix(urlPath, realPath)
+			accepted = true
+			return
+		},
+		Ready: func() error {
+			select {
+			case <-readyCh:
+				return nil
+			default:
+				return errors.New("syncer virtual workspace controllers are not started")
+			}
+		},
+		BootstrapAPISetManagement: func(mainConfig genericapiserver.CompletedConfig) (apidefinition.APIDefinitionSetGetter, error) {
+
+			clusterInformer := wildcardKcpInformers.Workload().V1alpha1().WorkloadClusters().Informer()
+			apiResourceImportInformer := wildcardKcpInformers.Apiresource().V1alpha1().APIResourceImports()
+			negotiatedAPIResourceInformer := wildcardKcpInformers.Apiresource().V1alpha1().NegotiatedAPIResources()
+
+			installedAPIs = newInstalledAPIs(func(logicalClusterName logicalcluster.Name, spec *apiresourcev1alpha1.CommonAPIResourceSpec) (apidefinition.APIDefinition, error) {
+				return apiserver.CreateServingInfoFor(mainConfig, logicalClusterName, spec, provideForwardingRestStorage(&clusterAwareClientGetter{
+					clusterInterface: dynamicClusterClient,
+				}))
+			})
+
+			// This should be replace by a real controller when the URLs should be added to the WorkloadCluster object
+			clusterInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					if workloadCluster, ok := obj.(*workloadv1alpha1.WorkloadCluster); ok {
+						workloadClusterRef := syncer.WorkloadClusterRef{
+							LogicalClusterName: logicalcluster.From(workloadCluster),
+							Name:               workloadCluster.Name,
+						}
+						installedAPIs.addWorkloadCluster(workloadClusterRef)
+					}
+				},
+				DeleteFunc: func(obj interface{}) {
+					if workloadCluster, ok := obj.(*workloadv1alpha1.WorkloadCluster); ok {
```

Comment:
this is not enough. It can be a tombstone on informer restart. We have to extract the key and decompose it.

Reply:
noticed that we have this issue also in other controllers. Fixing.

Reply:
done